src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -1120,6 +1131,89 @@ get_rebalance_progress(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * BuildShardSizesHash creates a hash that maps a shardid to its full size
+ * within the cluster. It does this by using the rebalance progress monitor
+ * state to find the node the shard is currently on. It then looks up the shard
+ * size in the shardStatistics hashmap for this node.
+ */
+static HTAB *
+BuildShardSizesHash(ProgressMonitorData *monitor, HTAB *shardStatistics)
+{
+	HASHCTL info = {
+		.keysize = sizeof(uint64),
+		.entrysize = sizeof(ShardStatistics),
+		.hcxt = CurrentMemoryContext
+	};
+
+	HTAB *shardSizes = hash_create(
+		"ShardSizeHash", 32, &info,
+		HASH_ELEM | HASH_CONTEXT | HASH_BLOBS);
+	PlacementUpdateEventProgress *placementUpdateEvents = monitor->steps;
+	for (int eventIndex = 0; eventIndex < monitor->stepCount; eventIndex++)
+	{
+		PlacementUpdateEventProgress *step = placementUpdateEvents + eventIndex;
+		uint64 shardId = step->shardId;
+		uint64 shardSize = 0;
+		uint64 backupShardSize = 0;
+
+		uint64 sourceSize = WorkerShardSize(shardStatistics, step->sourceName,
+											step->sourcePort, shardId);
+		uint64 targetSize = WorkerShardSize(shardStatistics, step->targetName,
+											step->targetPort, shardId);
+
+
+		if (step->progress == REBALANCE_PROGRESS_WAITING ||
+			step->progress == REBALANCE_PROGRESS_MOVING)
+		{
+			/*
+			 * If we are not done with the move, the correct shard size is the
+			 * size on the source.
+			 */
+			shardSize = sourceSize;
+			backupShardSize = targetSize;
+		}
+		else if (step->progress == REBALANCE_PROGRESS_MOVED)
+		{
+			/*
+			 * If we are done with the move, the correct shard size is the size
+			 * on the target
+			 */
+			shardSize = targetSize;
+			backupShardSize = sourceSize;
+		}
+
+		if (shardSize == 0)
+		{
+			if (backupShardSize == 0)
+			{
+				/*
+				 * We don't have any useful shard size. This can happen when a
+				 * shard is moved multiple times and it is not present on
+				 * either of these nodes. Probably the shard is on a worker
+				 * related to another event. In the weird case that this shard
+				 * is on the nodes and actually is size 0, we will have no
+				 * entry in the hashmap. When fetching from it we always
+				 * default to 0 if no entry is found, so that's fine.
+				 */
+				continue;
+			}
+
+			/*
+			 * If the shard size was 0, but the unexpected shard size.
```

Comment:
This comment is a bit sparse. I guess we can do `s/but/use`, or add `is not` at the end of the last sentence.